backend/core/src/listings/listings.service.ts

```diff
@@ -38,29 +48,61 @@ export class ListingsService {
       .leftJoinAndSelect("listings.reservedCommunityType", "reservedCommunityType")
   }
 
-  public async list(
-    origin: string,
-    jsonpath?: string,
-    filter?: ListingFilterParams[]
-  ): Promise<Listing[]> {
-    const qb = this.getQueryBuilder()
-    if (filter) {
-      addFilter<ListingFilterParams>(filter, "listings", qb)
+  public async list(origin: string, params: ListingsQueryParams): Promise<PaginatedListingsDto> {
+    // Inner query to get the sorted listing ids of the listings to display
+    // TODO(avaleske): Only join the tables we need for the filters that are applied
+    const innerFilteredQuery = this.listingRepository
+      .createQueryBuilder("listings")
+      .select("listings.id", "listings_id")
+      .leftJoin("listings.property", "property")
+      .groupBy("listings.id")
+      .orderBy({ "listings.id": "DESC" })
+
+    if (params.filter) {
+      addFilters<ListingFilterParams, typeof filterTypeToFieldMap>(
+        params.filter,
+        filterTypeToFieldMap,
+        innerFilteredQuery
+      )
     }
 
-    qb.orderBy({
-      "listings.id": "DESC",
-      "units.max_occupancy": "ASC",
-      "preferences.ordinal": "ASC",
-    })
+    const paginate = params.limit !== "all" && params.limit > 0 && params.page > 0
+    if (paginate) {
+      // Calculate the number of listings to skip (because they belong to lower page numbers).
+      const offset = (params.page - 1) * (params.limit as number)
```

Comment:
If the validation is changed to allow string values other than "all", this type assertion would cause an error. I'm open to suggestions for how to type guard this, since TypeScript isn't clever enough to realize that the type is narrowed by the `paginate` variable.

Reply:
This will work with typescript 4.4, but for now I don't have any ideas for you.

Reply:
Ah good to know! I added a todo explaining when we can remove the extra assertions.